src/buildercore/bootstrap.py

```diff
@@ -77,22 +78,27 @@ def create_stack(stackname):
     on_start = _noop
     on_error = _noop
     if pdata['aws']['ec2']:
-        parameters.append(('KeyName', stackname))
+        parameters.append({'ParameterKey': 'KeyName', 'ParameterValue': stackname})
         on_start = lambda: keypair.create_keypair(stackname)
         on_error = lambda: keypair.delete_keypair(stackname)
 
     return _create_generic_stack(stackname, parameters, on_start, on_error)
 
 def _create_generic_stack(stackname, parameters=None, on_start=_noop, on_error=_noop):
     "transforms templates stored on the filesystem into real resources (AWS via CloudFormation, Terraform)"
+    if core.stack_is_active(stackname):
+        LOG.info("stack exists") # avoid on_start handler
```

Comment:
I assumed the code shouldn't get here if the stack already exists?

Reply:
But idempotence is welcome

Reply:
it does actually, the only thing stopping it is a check in `cfn.launch`. without it, the `on_start` succeeds the first time in creating the keypair but fails the next time with a 'file exists' type error.